### PR TITLE
Fix missing header rejection for custom headers #2487

### PR DESCRIPTION
**Pull request by mfulgo**
## Purpose
When the HeaderMagnet is formed for a user-defined custom header, it
used the class' name to populate the `headerName` field. However, the
user may have named their class differently than its name, by excluding
dashes, for example.

This could produce an inconsistency in behavior when routes would
generate a `MissingHeaderRejection` for a custom header because the
`headerValueByType` directive constructs a `MissingHeaderRejection` with
a `headerName` populated from the magnet, which in the case of a custom
header, comes from the header's class name rather than its name field.

## References
References #2487 

## Changes
* Updates the `HeaderMagnet` implicit for `ModeledCustomHeader`s to set the `headerName` from the companion's `name` field.

## Background Context
* See #2487 for background and an example of the undesired behavior.
* This approach was taken because it's a focused, one-line change that should have a fairly limited effect. See 1bcc63e for an alternative approach that was considered.

### CODE REVIEW

**Comment by jrudolph** [comment]
OK TO TEST

**Comment by mfulgo** [comment]
@jrudolph @johanandren - I'm not sure what the code review process or timeline is, but this should be ready for one.

**Comment by johanandren** [comment]
Thanks @mfulgo, we have been traveling for team meetings so please bear with us waiting for reviews.

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/HeaderDirectivesSpec.scala [hunk 1]
```diff
@@ -202,4 +206,30 @@ class HeaderDirectivesSpec extends RoutingSpec with Inside {
       }
     }
   }
+
+  "The HeaderMagnet" should {
```
Comment (jrudolph) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/HeaderDirectivesSpec.scala]: HeaderMagnet is only an implementation detail, so not sure if we need exactly those tests. Could you additionally (or instead) create one or more tests, e.g. above for `headerValueForType` that asserts whether the rejection is returned similar to the test you changed in the ModeledCustomHeaderSpec?

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala [hunk 2]
```diff
@@ -199,7 +200,9 @@ trait LowPriorityHeaderMagnetImplicits {
     new HeaderMagnet[T] {
       override def classTag: ClassTag[T] = ClassTag(clazz)
       override def runtimeClass: Class[T] = clazz
-      override def extractPF: PartialFunction[HttpHeader, T] = { case x if runtimeClass.isAssignableFrom(x.getClass) ⇒ x.asInstanceOf[T] }
+      override def extractPF: PartialFunction[HttpHeader, T] = {
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala]: Let's keep the original formatting here and below.

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala [hunk 3]
```diff
@@ -181,11 +181,12 @@ object HeaderMagnet extends LowPriorityHeaderMagnetImplicits {
...
     new HeaderMagnet[T] {
-      override def runtimeClass = tag.runtimeClass.asInstanceOf[Class[T]]
-      override def classTag = tag
-      override def extractPF = {
+      override def classTag: ClassTag[T] = tag
+      override def runtimeClass: Class[T] = tag.runtimeClass.asInstanceOf[Class[T]]
+      override def extractPF: PartialFunction[HttpHeader, T] = {
         case h if h.is(companion.lowercaseName) ⇒ companion.apply(h.value)
       }
+      override def headerName: String = companion.name
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala]: :+1:

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/HeaderDirectivesSpec.scala (hunk 1 above)
Comment (mfulgo) [reply akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/HeaderDirectivesSpec.scala]: Yeah. That should be easy enough. I'll do that when I get some free time, maybe later this week.

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala (hunk 2 above)
Comment (mfulgo) [reply akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala]: Line 187 was on a second line, so the intention here was to give it some consistency across the class and hopefully make it a little more readable. Let me know if it's still not what you want.

**Comment by mfulgo** [comment]
@jrudolph Added the test you asked for. Let me know about the other comment.

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala (hunk 2 above)
Comment (jrudolph) [reply akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala]: Ah, I see. Sounds good, then.

**Pull request merged**